src/main.c

```diff
@@ -3310,15 +3307,26 @@ int main(int argc, char **argv)
 
     /* start the threads */
     conf.threads = alloca(sizeof(conf.threads[0]) * conf.thread_map.size);
-    size_t i;
-    for (i = 1; i != conf.thread_map.size; ++i) {
-        pthread_t tid;
-        h2o_multithread_create_thread(&tid, NULL, run_loop, (void *)i);
-    }
+    pthread_t *tids = alloca(sizeof(*tids) * conf.thread_map.size);
+    for (size_t i = 1; i != conf.thread_map.size; ++i)
+        h2o_multithread_create_thread(&tids[i], NULL, run_loop, (void *)i);
 
     /* this thread becomes the first thread */
     run_loop((void *)0);
 
-    /* notreached */
+    /* wait for all threads to exit */
+    for (size_t i = 1; i != conf.thread_map.size; ++i) {
+        while (pthread_join(tids[i], NULL) != 0) {
+            if (errno != EINTR) {
```

Comment:
Is `pthread_join` supposed to set `errno` on `EINTR`? By reading [pthread_join(3) manual](https://man7.org/linux/man-pages/man3/pthread_join.3.html) it is implied that an error code would be returned as a return value, though `EINTR` is not mentioned as a possible return value. So while not 100% sure my recommendation would be to compare the return value (not `errno`) against `EINTR`.

It looks like that the [latest POSIX (2016 or later?) disallowed `EINTR` as a return value of `pthread_join`](https://pubs.opengroup.org/onlinepubs/9699919799.2016edition/xrat/V4_xsh_chap02.html) (see section "Disallowing Return of the [EINTR] Error Code"). I'm not sure how many of libc implementations are up to this latest standard. At least comparing the return value against `EINTR` should be harmless, even if the implementation adopts this new standard (the condition would never be met in this case).

Reply:
Thank you for the review and the research. This is very informative.

I was trying to be safe, but reading your comment and links, I think we should remove this. We do not check for `EINTR` when using other pthread calls (e.g., `pthread_mutex_lock`), doing it just here does not make sense.

Let's call `perror` followed by an `abort`, whenever `pthread_join` fails. Then, if we see `EINTR` on some platform, we can consider if use of our pthread library calls have to be adjusted, not just here, but everywhere.

Reply:
Addressed in dd09a69.